Fix OCB partial block issue

mattcaswell opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->

This fixes a bug in OCB where if an "Update" function is called with a partial block of data, then a subsequent "Update" will use the partial data already supplied and write out the block as expected, but then overwrite it with the next full block of output.

I also updated the tests to catch this case.

Fixes #2273

mattcaswell commented:
Writing the test for this threw up a few things. Firstly XTS and wrap do not seem to be able to cope with unusual (partial block) input lengths. This is possibly expected as these modes are "special".

I also found that CCM also could not handle unusual lengths. I am less sure whether this is expected behaviour or not.

Perhaps @snhenson can shed some light on the above?

I also discovered that the `is_partially_overlapping()` check fails if you supply partial blocks of data to an update function whilst using in place encryption. This looks like a possible bug to me. @dot-asm your view?

dot-asm commented:
Problem with suggested evp_test.c modification is that there is a number of vectors in evptests.txt with lengths specifically chosen to exercise corner cases [customarily] in vector code paths. And if you chip off bytes for EVP to handle, some of intended code paths won't be exercised. Well, as of now they will be exercised in in-place case, but one can argue that other cases are relevant too. What I'm trying to say is that fragmented input should probably be *additional* separate test, not replacement...

richsalz commented on crypto/evp/e_aes.c:
```diff
@@ -2574,21 +2574,25 @@ static int aes_ocb_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
+            written_len = AES_BLOCK_SIZE;
             *buf_len = 0;
+            if (out != NULL)
+                out+= AES_BLOCK_SIZE;
```
space before the operator.  yes, a real obsessive-compulsive behavior kind of nit, but this is about OCB innit?

mattcaswell replied:
Fixed.

richsalz commented on test/evp_test.c:
```diff
@@ -983,15 +983,67 @@ static int cipher_test_enc(struct evp_test *t, int enc,
+            if (cdat->aad_len > 0) {
+                if (!EVP_CipherUpdate(ctx, NULL, &chunklen, cdat->aad, 1))
+                    goto err;
+                donelen += 1;
```
i get the += for parallelism, but it does jangle to not see ++

mattcaswell replied:
Changed.

richsalz commented on crypto/evp/e_aes.c:
```diff
@@ -2557,10 +2557,10 @@ static int aes_ocb_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
             unsigned int remaining;
 
-            remaining = 16 - (*buf_len);
+            remaining = AES_BLOCK_SIZE - (*buf_len);
```
yeah for using defines!

kroeckx commented:
It looks broken for in_len == 3 and aad_len == 3

mattcaswell commented:
> It looks broken for in_len == 3 and aad_len == 3

How? I don't see it??

mattcaswell commented:
> Problem with suggested evp_test.c modification is that there is a number of vectors in evptests.txt with lengths specifically chosen to exercise corner cases [customarily] in vector code paths. And if you chip off bytes for EVP to handle, some of intended code paths won't be exercised. Well, as of now they will be exercised in in-place case, but one can argue that other cases are relevant too. What I'm trying to say is that fragmented input should probably be additional separate test, not replacement...

I changed things around a little so that it is now an additional test rather than a replacement.

mattcaswell commented:
New commits pushed addressing all issues.

I also added a new commit to deal with the issue of `is_partially_overlapping()` failing when using fragmented input and in place encryption. Perhaps @dot-asm might want to take a look.

kroeckx commented:
On Tue, Jan 24, 2017 at 05:04:08AM -0800, mattcaswell wrote:
> > It looks broken for in_len == 3 and aad_len == 3
> 
> How? I don't see it??

I got confused. In any case, it's hard to follow code. Please
consider adjusting in and in_len instead of using donelen.

mattcaswell commented:
> I got confused. In any case, it's hard to follow code. Please
consider adjusting in and in_len instead of using donelen.

I added an extra commit to do that - although it only makes sense in the non-aad portion of the code (the aad portion uses non-local variables instead of `in` and `in_len` so it doesn't make sense to change them).

mattcaswell commented:
I'd really like to get this in the release tomorrow if someone is able to review.

richsalz commented on crypto/evp/evp_enc.c:
```diff
@@ -276,8 +276,7 @@ int EVP_DecryptInit_ex(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
 # define PTRDIFF_T size_t
 #endif
 
-static int is_partially_overlapping(const void *ptr1, const void *ptr2,
```
this could be called ```overlaps```.

richsalz commented:
I think the assert() in is_partially_overlapping is now completely wrong. Look at the places it's used.

richsalz commented:
I had already plus-one'd this.  Fix the assert issue and I reconfirm.

mattcaswell commented:
assert removed. I will merge shortly. And yes this is for master and 1.1.0 as per your earlier comment.

mattcaswell commented:
Pushed. Thanks.
